### ISSUE TITLE
Type mismatch for two identical Union[*Ts] variables

### ISSUE THREAD

**Issue by toinbis** [microsoft/pyright#5775]
**Describe the bug**
Type mismatch for two identical Union[*Ts] variables gives `type "Union[*Ts@X]" cannot be assigned to parameter "__object" of type "*TS@Y"` error, though the code seems to be valid.

**Code or Screenshots**

```python
from typing import Callable, Union, TypeVarTuple, Generic

class Dog:
    pass

class Cat:
    pass

class MyFirstDog(Dog):
    pass

class MyFirstCat(Cat):
    pass

class MySecondDog(Dog):
    pass

class MySecondCat(Cat):
    pass

HandlerType1 = Callable[[MyFirstDog, MyFirstCat], tuple[MyFirstDog, MyFirstCat]]
HandlerType2 = Callable[[MySecondDog, MySecondCat], tuple[MySecondDog, MySecondCat]]

HandlerTypes = TypeVarTuple('HandlerTypes')

class AppModule(Generic[*HandlerTypes]):
    def __init__(self) -> None:
        self.handlers: list[Union[*HandlerTypes]] = []

    def pet_handler(self, handler_func: Union[*HandlerTypes]) -> Union[*HandlerTypes]:
        self.handlers.append(handler_func)
        return handler_func

appmodule = AppModule[HandlerType1, HandlerType2]()

@appmodule.pet_handler
def pet_handler1(a: MyFirstDog, b: MyFirstCat) -> tuple[MyFirstDog, MyFirstCat]:
    return (a, b)

@appmodule.pet_handler
def pet_handler2(a: MySecondDog, b: MySecondCat) -> tuple[MySecondDog, MySecondCat]:
    return (a, b)

```
`self.handlers.append(handler_func)` of the `pet_handler` method gives `Argument of type "Union[*HandlerTypes@AppModule]" cannot be assigned to parameter "__object" of type "*HandlerTypes@AppModule" in function "append"
  Type "Union[*HandlerTypes@AppModule]" cannot be assigned to type "*HandlerTypes@AppModule"Pylance[reportGeneralTypeIssues](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportGeneralTypeIssues)` error.

**VS Code extension or command-line**
Pylance VSCode(Insiders build) extension v2023.8.31 Pre-release.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.324, which I just published. It will also be included in a future release of pylance.
